### PR TITLE
More respectful convention for caching

### LINKED ISSUE #342: Caching: Use a more respectful default location?
At the moment, we fall back to `~/.pip-audit/cache` if `pip cache dir` fails. This isn't a "respectful" default, since it doesn't adhere to either XDG (which probably wants us to consult `XDG_CACHE_HOME`) or macOS's conventions (`~/Library/Caches`, probably).

This probably isn't the most important change to make since 99% of users won't notice it anyways, but it would make us a slightly better participant in the user's home directory.

### PR DESCRIPTION
**Pull request by aanghelidi**
A PR that aims to close #342.

All the logic is inside `_get_internal_cache_path` in `_cache.py`.

### CODE REVIEW

**Review by woodruffw — APPROVED** [review]
Thanks @aanghelidi, this looks pretty reasonable to me. `pip-audit`'s choice of cache location isn't a stability commitment we make, so we should be fine to change this without making a major release.

A couple of thoughts:

* Maybe we could use [`platformdirs`](https://pypi.org/project/platformdirs/) to abstract the cache location discovery for us? OTOH the current logic in `_get_internal_cache_path` is pretty simple; curious what you think.
* Maybe we should blow away the old cache if a new path is discovered? Otherwise, users will end up silently keeping `_PIP_AUDIT_INTERNAL_CACHE` around in their `$HOME` taking up space, which isn't very polite of us.

For the second point, something like this is what I have in mind:

```python
# if the selected internal cache isn't the old one, delete the old one.
if internal_cache_path != _PIP_AUDIT_LEGACY_INTERNAL_CACHE:
    shutil.rmtree(_PIP_AUDIT_LEGACY_INTERNAL_CACHE)
```

CC @di for thoughts as well 🙂

**Comment by aanghelidi** [comment]
Thanks @woodruffw  for the review 🙂 TIL about [platformdirs](https://pypi.org/project/platformdirs/) ! This is indeed a really handy package. Regarding the aforementioned points:

- There is several pros about using `platformdirs`:
  - simplicity. Using `platformdirs`, the implementation of `pip-audit`'s cache location is a one-liner.
  - we do not reinvent the wheel. This means less maintenance for the project.
  - consistency. Upon research, it seems that `platformdirs` is already used in several `pypa` projects:
    -  [hatch](https://github.com/pypa/hatch/tree/master): see [here](https://github.com/pypa/hatch/blob/e07d2984aba528262616b70891860e5aefb12775/src/hatch/config/model.py#L414)
    - [pipx](https://github.com/pypa/pipx/tree/main): see [here](https://github.com/pypa/pipx/blob/4462299fa391b7349074facc39c73767853c9d02/src/pipx/paths.py#L118)
    - [setuptools](https://github.com/pypa/setuptools): see [here](https://github.com/pypa/setuptools/tree/main/setuptools/_vendor/platformdirs) (as a vendored library)

- Regarding the potential con(s):
  - dependency cost. We add one more dependency to the project. Although `platformdirs` itself has no dependencies, it's still a decision to make. Maybe this is not worth adding given the simplicity of `_get_internal_cache_path` implementation (which I have the feeling could be simplified).
  
- Regarding `PIP_AUDIT_INTERNAL_CACHE`, it is indeed not nice to keep the old cache in `$HOME`. Removing it seems to be a good idea. Maybe, as an alternative, using `shutil.move` to move the old cache to the new location could be an idea?

I added a commit using `platformdirs`  for the cache location, let me know what you think 🙂

**Comment by woodruffw** [comment]
> * Regarding `PIP_AUDIT_INTERNAL_CACHE`, it is indeed not nice to keep the old cache in `$HOME`. Removing it seems to be a good idea. Maybe, as an alternative, using `shutil.move` to move the old cache to the new location could be an idea?

I think we should just blow it away for now -- #794 covers some larger architectural changes we need to make to handle newer versions of `pip`'s cache anyways, so these old paths may not be super useful when moved to the new location.

> I added a commit using platformdirs for the cache location, let me know what you think 🙂

LGTM, thank you!

**Comment by aanghelidi** [comment]
> I think we should just blow it away for now -- https://github.com/pypa/pip-audit/issues/794 covers some larger architectural changes we need to make to handle newer versions of pip's cache anyways, so these old paths may not be super useful when moved to the new location

Thank you for the context @woodruffw ! I added a extra commit to delete the old cache that goes into that direction. Let me know what you think 🙂

File: pip_audit/_cache.py [hunk 1]
```diff
@@ -46,6 +48,14 @@ def _get_pip_cache() -> Path:
     return http_cache_dir
 
 
+def _delete_legacy_cache_dir(current_cache_dir: Path, legacy_cache_dir: Path) -> None:
+    """
+    Deletes the legacy `pip-audit` if it exists.
+    """
+    if current_cache_dir != legacy_cache_dir:
+        shutil.rmtree(legacy_cache_dir)
```
Comment (woodruffw) [inline pip_audit/_cache.py:56]: No need for a helper here; let's just inline this into `_get_cache_dir` IMO.

Comment (aanghelidi) [reply pip_audit/_cache.py]: The last commit now inline the deletion into `_get_cache_dir` :heavy_check_mark:

**Review by woodruffw — APPROVED** [review]
Thanks @aanghelidi, LGTM!

**Review by woodruffw — COMMENTED** [review]
Sorry, one last request: could you add a `CHANGELOG` entry for this change? See the other entries in the log for our preferred style.

**Review by woodruffw — APPROVED** [review]

**Comment by hseg** [comment]
Tests fails on my machine -- with `$XDG_CACHE_HOME=~/.local/var/cache/`, I get
```
WARNING  pip_audit._cache:_cache.py:81 pip 1.0.0 doesn't support the `cache dir` subcommand, using /home/gesh/.local/var/cache/pip-audit instead
FAILED test/test_cache.py::test_get_cache_dir_do_not_use_pip[on Linux] - AssertionError: assert PosixPath('/home/gesh/.local/var/cache/pip-audit') == PosixPath('/home/gesh/.c...
FAILED test/test_cache.py::test_get_cache_dir_pip_disabled_in_environment[on Linux] - AssertionError: assert PosixPath('/home/gesh/.local/var/cache/pip-audit') == PosixPath('/home/gesh/.c...
FAILED test/test_cache.py::test_get_cache_dir_old_pip[on Linux] - AssertionError: assert PosixPath('/home/gesh/.local/var/cache/pip-audit') == PosixPath('/home/gesh/.c...
```
which is odd since I have
```
$ pip --version
pip 25.0 from /usr/lib/python3.13/site-packages/pip (python 3.13)
```

**Comment by hseg** [comment]
Judging by the log output and the testcases, I suspect what's going on is you've
correctly implemented the "check XDG variables and fallback" logic, but are not
accounting for it in the test cases.
In that case, it'd be enough to unset `$XDG_CACHE_HOME` in the three
above-mentioned tests (and add a test exercising the expected behaviour that if
`$XDG_CACHE_HOME` is set but `pip` cannot be relied on to give the path the
obtained cache directory is indeed `$XDG_CACHE_HOME/pip-audit` -- perhaps by
setting `$XDG_CACHE_HOME` to a canary value)

**Pull request merged**